lib/ukfs/include/uk/fs.h

```diff
@@ -0,0 +1,591 @@
+
+/**
+ * Graft the directory `f` into the filesystem of `ref` in its position.
+ *
+ * Grafting is the complementary, yet orthogonal, counterpart to mounting, being
+ * a runtime-volatile property of a filesystem instance that influences the
+ * behavior of lookup: lookups of `f`'s parent should instead return the parent
+ * of `ref`.
+ *
+ * This function is guaranteed to be called at most once on a directory
+ * instance, and not concurrent with other operations.
+ * Typically this function is called on the root of a new (bind) mount when it
+ * is being attached to the filesystem.
+ *
+ * This operation may fail with -ENOSYS if not supported by the driver.
+ *
+ * @param f Directory to graft
+ * @param ref Graft point; directory to delegate parent lookups to
+ *
+ * @return
+ *  == 0: Success
+ *   < 0: Negative error code on failure
+ */
+typedef int (*uk_fs_graft_func)(const struct uk_file *f,
+				const struct uk_file *ref);
+
+/**
+ * Filesystem path, length included; used as return value
+ */
+struct uk_fs_path {
+	const char *s; /* NUL-terminated string containing path */
+	size_t len; /* Length of string, same as returned by strlen(.s) */
+};
+
+/**
+ * Get the path pointed to by symlink `f` as efficiently as possible.
+ *
+ * The returned string is guaranteed a lifetime at least as long as `f`.
+ * This operation should avoid any memory allocation or copying.
+ * Callers must treat returned memory as immutable.
+ * This operation cannot fail.
+ *
+ * @param f Symlink to read
+ *
+ * @return
+ *  Path that the symlink points to.
+ */
+typedef struct uk_fs_path (*uk_fs_readlink_func)(const struct uk_file *f);
+
+/* Filesystem ops table */
+struct uk_fs_ops {
+	/* Ops targeting the underlying filesystem itself */
+	uk_fs_stat_func stat;
+	uk_fs_sync_func sync;
+	/* Ops targeting individual files */
+	/* Universal ops; must be implemented by all filesystem-backed files */
+	uk_fs_lookup_func lookup;
+	uk_fs_mount_func mount;
+	uk_fs_rebind_func rebind;
+	/* Directory ops; must be NULL for non-directories */
+	uk_fs_listdir_func listdir;
+	uk_fs_create_func create;
+	uk_fs_unlink_func unlink;
+	uk_fs_rename_func rename;
+	uk_fs_graft_func graft;
+	/* Symlink ops; must be NULL for non-symlinks */
+	uk_fs_readlink_func readlink;
+	/* Constraints on open file mode
+	 * These bits will always be set on open files originating from this FS
+	 */
+	unsigned int constraints;
+};
+
+/* Chosen to map to similar consts in posix-fd for simplicity & performance */
+/* File does not support seeking */
+#define UKFS_MODE_NOSEEK   010
+/* Driver handles synchronization internally */
+#define UKFS_MODE_NOIOLOCK 020
+
+/* Only filesystem nodes have `fsops` set */
+#define UKFS_FILE_ISFS(f) (!!(f)->fsops)
+/* Non-directories must not implement `listdir` */
+#define UKFS_FILE_ISDIR(f) (UKFS_FILE_ISFS(f) && (f)->fsops->listdir)
+/* Only symlinks may implement `readlink` */
+#define UKFS_FILE_ISSYM(f) (UKFS_FILE_ISFS(f) && (f)->fsops->readlink)
```

Comment:
I see multi-eval macro again where someone could pass a function call as an argument. Would it be worth making it a static inline instead?